### PR TITLE
feat: store event-type filters to localStorage

### LINKED ISSUE #13394: [CAL-3015] event-type filters dont persist
we should store the setting here in localStorage

CleanShot 2024-01-24 at 18 08 23@2x

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-3015](https://linear.app/calcom/issue/CAL-3015/event-type-filters-dont-persist)</sub>

### PR DESCRIPTION
**Pull request by abhijeetsingh-22**
## What does this PR do?
This PR introduces a new feature that stores `event-type` filter values in localStorage, ensuring that the selected filter persists when users reload the page.

Fixes #13394 

https://github.com/calcom/cal.com/assets/55000107/228f2be2-7139-471a-ac2a-81b2440f1aea

## Type of change

- New feature (non-breaking change which adds functionality)

## How should this be tested?
- Open `/event-types`.
- Change the event-type filter .
- Reload the page. 

### Desired Outcome:
 The event-type filter you selected previously should persist, and the query parameters should be appropriately populated.

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist
All done

### CODE REVIEW

File: packages/lib/hooks/useTypedQuery.ts
Comment (supalarry) [inline packages/lib/hooks/useTypedQuery.ts:39]: How about we remove `useLocalStorage` key and only rely on `localStorageKey` - if `localStorageKey` is not passed (can set it's default value to null) then `useLocalStorage` would be always false. If developer passes `localStorageKey` then it implicitly implies that developer wants to use local storage. So I think its better to have for simplicity:

```
export function useTypedQuery<T extends z.AnyZodObject>(
  schema: T,
  localStorageKey = null
) {
```

File: packages/features/filters/components/TeamsFilter.tsx
Comment (supalarry) [inline packages/features/filters/components/TeamsFilter.tsx:20]: To reduce possibility that "TEAMS_FILTER_KEY" is used elsewhere I would append "EVENT_TYPES_TEAMS_FILTER_KEY" in front to reduce possibility of collision but its your call.

**Review by supalarry — COMMENTED** [review]
Great work as always Abhijeet! I tested PR locally and it works.

I have 1 suggestion on simplifying the usage of "useTypedQuery" by passing only 2 parameters. After that it is good.

File: packages/lib/hooks/useTypedQuery.ts
Comment (supalarry) [reply packages/lib/hooks/useTypedQuery.ts:39]: What do you think?

Comment (abhijeetsingh-22) [reply packages/lib/hooks/useTypedQuery.ts:39]: @supalarry thanks for the review. You are right we should remove the redundant parameter. I will make the changes and push

File: packages/features/filters/components/TeamsFilter.tsx
Comment (abhijeetsingh-22) [reply packages/features/filters/components/TeamsFilter.tsx:20]: Sure will also change this

File: packages/lib/hooks/useTypedQuery.ts
Comment (supalarry) [reply packages/lib/hooks/useTypedQuery.ts:39]: Amazing!

File: packages/features/filters/components/TeamsFilter.tsx
Comment (supalarry) [reply packages/features/filters/components/TeamsFilter.tsx:20]: Yeah actually now that i think about it the current implementation is good - i mistakenly thought the file was in event-types related folder.

File: packages/lib/hooks/useTypedQuery.ts
Comment (PeerRich) [reply packages/lib/hooks/useTypedQuery.ts:39]: awesome thank you! will approve after the change

**Comment by abhijeetsingh-22** [comment]
Made the requested changes

**Comment by supalarry** [comment]
> Made the requested changes

Amazing, looks good now! 🚀

**Review by supalarry — APPROVED** [review]

**Pull request merged**
